fix: validate empty resources for non-suffix responders

sonephyo opened the pull request:
# Summary of Changes

This PR fixes suffixed and normal responders to have consistent behavior by raising MethodNotFound when method_map is empty. This is to disable allowing empty resources by nature, as mentioned in Issue #2167. 

Test cases related to Stonewall are an empty Resource, which contradicts, and therefore, is removed.
Test case test_ws_simulator_collect_edge_cases is updated to include on_websocket function.
Test case (test_custom_error_route_not_found) is introduced for testing an empty resource.

# Related Issues

Closes #2167 

# Pull Request Checklist

This is just a reminder about the most common mistakes.  Please make sure that you tick all *appropriate* boxes.  Reading our [contribution guide](https://falcon.readthedocs.io/en/stable/community/contributing.html) at least once will save you a few review cycles!

If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing to do.

- [x] Applied changes to both WSGI and ASGI code paths and interfaces (where applicable).
- [x] Added **tests** for changed code.
- [x] Performed automated tests and code quality checks by [running `tox`](https://falcon.readthedocs.io/en/stable/community/contributing.html#pull-requests).
- [x] Prefixed code comments with GitHub nick and an appropriate prefix.
- [x] Coding style is consistent with the rest of the framework.
- [x] Updated **documentation** for changed code.
    - [x] Added docstrings for any new classes, functions, or modules.
    - [x] Updated docstrings for any modifications to existing code.
    - [x] Updated both WSGI and ASGI docs (where applicable).
    - [x] Added references to new classes, functions, or modules to the relevant RST file under `docs/`.
    - [x] Updated all relevant supporting documentation files under `docs/`.
    - [x] A copyright notice is included at the top of any new modules (using your own name or the name of your organization).
    - [x] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/usage/restructuredtext/directives.html?highlight=versionadded#directive-versionadded).
- [x] Changes (and possible deprecations) have [towncrier news fragments](https://falcon.readthedocs.io/en/stable/community/contributing.html#changelog) under `docs/_newsfragments/`, with the file name format `{issue_number}.{fragment_type}.rst`. (Run `tox -e towncrier`, and inspect `docs/_build/html/changes/` in the browser to ensure it renders correctly.)
- [x] LLM output, if any, has been carefully reviewed and tested by a human developer. (See also: [Use of LLMs ("AI")](https://falcon.readthedocs.io/en/latest/community/contributing.html#use-of-llms-ai).)

sonephyo committed:
fix: raise error for responders not found for resources without suffix

- Change from SuffixedMethodNotFoundError to MethodNotFoundError
- Added Indicators for the resource and suffix

sonephyo committed:
test: add test_custom_error_route_not_found test

sonephyo committed:
test: remove redundant stonewall setup test cases

- Stonewall related test cases removed
- Already covered by test_custom_error_route_not_found which tests that empty resources raise MethodNotFoundError

sonephyo committed:
test: fix ws Resource to include on_websocket

vytas7 reviewed (changes requested):
Thanks for this contribution @sonephyo!

I think this looks good for most of it, however, Falcon is known for going to great lengths for avoiding breaking surprises. So I think we should just emit a warning for now, and actually start raising an error only in 5.0.

Also, we need a newsfragment to describe what has been improved or fixed.

CaselIT reviewed (commented):
great, but as mentioned let's start with warning

sonephyo committed:
docs(newsfragments): add rst with fixes

sonephyo committed:
fix: change non-suffix responders to emit a warning instead of error

MethodNotFoundError change back to SuffixedMethodNotFoundError
warnings.warn use emit UserWarning

sonephyo committed:
test: update test to verify UserWarning instead of MethodNotFoundError

sonephyo committed:
chore: run tox -e reformat

sonephyo committed:
Merge branch 'master' into bug/validate-empty-resources

vytas7 commented on tests/test_uri_templates.py:
```diff
@@ -602,3 +602,11 @@ def test_custom_error_on_suffix_route_not_found(client):
+    class EmptyResource:
+        pass
+
+    with pytest.warns(UserWarning):
```
If we leave the stonewall part, maybe this test becomes somewhat redundant. But, OTOH, it doesn't hurt.

vytas7 commented on tests/asgi/test_ws.py:
```diff
@@ -1087,7 +1087,8 @@ async def on_websocket(self, req, ws):
 async def test_ws_simulator_collect_edge_cases(conductor):
     class Resource:
-        pass
+        async def on_websocket(self, req, ws):
```
Maybe adding an unrelated method (such as `on_delete` or whatever) would match the original intent of this test case better? (I.e., not having any `on_websocket()` responder.)

vytas7 commented on docs/_newsfragments/2167.bugfix.rst:
```diff
@@ -0,0 +1,3 @@
+Resources without responder methods now emit a ``UserWarning`` for non-suffix responders.
```
We could expand the newsfragment a bit, telling a story what the inconsistency was, and how it was addressed.

vytas7 reviewed (changes requested):
Thanks, this looks good for most of it :100:

Just a couple of cosmetic nitpicks before merging (see inline).

vytas7 committed:
Merge branch 'master' into bug/validate-empty-resources

vytas7 commented:
Converting this to Draft until @sonephyo is back (because we have merge conflicts).
